### DISCUSSION TITLE
Wrong path at sub app with custom open api

### DISCUSSION THREAD

**Discussion by LorhanSohaky** [fastapi/fastapi#8614] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.openapi.utils import get_openapi
from fastapi.testclient import TestClient


app = FastAPI()
sub_app = FastAPI(
    docs_url=None,
    redoc_url=None,
    openapi_url=None,
)


@sub_app.get("/items/")
async def read_items():
    return {"id": "foo"}


@sub_app.get("/openapi.json")
async def openapi(
    # Custom Deps
):
    return get_openapi(
        title="Custom OpenAPI",
        version="0.1",
        routes=sub_app.routes,
        prefix="/sub_app",
    )

app.mount("/sub_app", sub_app)
```

### Description

Open API not including `/sub_app` in the paths

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

Python 3.7.12

**Comment by vitclarke279** [comment]
My work around was to modify the openapi schema, hope this helps someone :

```
def add_sub_app_prefix_to_routing(api_spec: dict, prefix: str) -> dict:
    """
    Add a specified prefix to all paths in the OpenAPI specification for routing.

    :param api_spec: Dictionary of the openapi schema.
    :type api_spec: dict
    :param prefix: The prefix to be added to all paths.
    :type prefix: string

    :returns: Modified api schema with the paths including the prefix.
    :rtype: dict
    """
    # Get the current paths
    current_paths = list(api_spec["paths"].keys())

    # Iterate over the current paths and modify the api_spec
    for path in current_paths:
        new_path = prefix + path
        api_spec["paths"][new_path] = api_spec["paths"].pop(path)

    return api_spec

def custom_openapi():
    openapi_schema = get_openapi(
        title="Some API",
        version="1.0.0",
        summary="API to get stuff",
        description="API to get stuff",
        routes=sub_api_app.routes,
    )

    fixed_open_api_schema = add_sub_app_prefix_routing(
        api_spec=openapi_schema, prefix="/sub_api"
    )

    sub_api_app.openapi_schema = fixed_open_api_schema
    return sub_api_app.openapi_schema


sub_api_app.openapi = custom_openapi
```

**Comment by YuriiMotov** [comment]
This is expected.
FastAPI treats these mounted apps as separate applications and shows paths relative to the app's mounting point.
Mounting path is added to `servers`, so client can construct correct URL using it


```json
  "servers": [
    {
      "url": "/sub_app"  // <- mounting point of sub-app
    }
  ],
  "paths": {
    "/items/": {  // <- path relative to the mounting point
```

**Comment by YuriiMotov** [comment, marked as answer]
The code example is outdated, but affter [this comment](https://github.com/fastapi/fastapi/pull/4657#issuecomment-3123374081) I looked at this problem one more time and I got it.

The `openapi.json` endpoint implementation in code example is simplified compare to default implementation.
If you look at the default implementation of `openapi.json` endpoint, you will see that before calling `get_openapi` (it's inside `self.openapi()`) it updates `servers` by adding `url` equal to `root_path`.

https://github.com/fastapi/fastapi/blob/d5d302c3da11262f4e320a9af5dbb9d13386e1b5/fastapi/applications.py#L999-L1011

The implementation shown in docs misses this part. It works well for apps that are mounted at `'/'`, but fails in other case.

So, to fix your problem you need to change the implementation of `/openapi.json` endpoint to something like:
```py
@private_subapp.get("/openapi.json")
async def openapi(
    credentials: Annotated[HTTPBasicCredentials, Depends(security)],
    req: Request,
):
    # The following 3 lines may be moved to lifespan event
    app = cast(FastAPI, req.app)
    urls = (server_data.get("url") for server_data in app.servers)
    server_urls = {url for url in urls if url}

    root_path = req.scope.get("root_path", "").rstrip("/")
    if root_path not in server_urls:
        if root_path and app.root_path_in_servers:
            app.servers.insert(0, {"url": root_path})
            server_urls.add(root_path)

    return get_openapi(
        title=f"PRIVATE: OpenAPI",
        version="0.1.0",
        routes=private_subapp.routes,
        servers=app.servers,  # <- added
    )
```

**Reply by LorhanSohaky** [reply]
Thanks for the suggestion, this was actually a very old code, at the time it was still version 0.75
